AJ-1633: Finish wiring up `deleteCollection`

jladieu opened the pull request:
For [AJ-1633](https://broadworkbench.atlassian.net/browse/AJ-1633), branch based off https://github.com/DataBiosphere/terra-workspace-data-service/pull/594

* Stop checking `delete` workspace permission to determine if it's ok to delete a collection, instead check `write` workspace.
* Eliminate redundant `SamAuthorizationDao` methods which are subsumed by `SamAuthorizationDao#hasWriteWorkspacePermission`
* Adjust `SamPactTest` to no longer verify `delete` interaction and instead verify previously untested `read` interaction.
* Adjust a bunch of tests which were verifying the safety check behavior which is now removed.
* This allows wiring up `deleteCollection` to correctly look up the `WorkspaceId` for the collection and then check for `write` permission.
* Add `tenancy.enforce-collections-match-workspace-id` which is enabled on `data-plane` but can be disabled for stubborn tests that violate the rule it is checking.

[AJ-1633]: https://broadworkbench.atlassian.net/browse/AJ-1633?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jladieu commented on service/src/main/java/org/databiosphere/workspacedataservice/service/CollectionService.java:
```diff
@@ -179,22 +176,6 @@ public WorkspaceId getWorkspaceId(CollectionId collectionId) {
       }
     }
-
-    // safety check: if we found a workspace id in the db, it indicates we are in a data-plane
```
Is it safe to remove this safety check yet, or do you think we need to put in more safeguards first?  My hunch is that since we're blocking the `@Deprecated` version of `createCollection` by disallowing creates for virtual collections & missing `$WORKSPACE_ID` that this should be safe now.

ashanhol replied:
is there an observability way for us to check how often we'd be hitting the `CollectionException`? If so, why not keep the safety check in and monitor for a bit until we're sure?

jladieu replied:
Added a kill switch which I only needed to use sparingly :shrug:

ashanhol commented on service/src/test/java/org/databiosphere/workspacedataservice/service/CollectionServiceGetWorkspaceIdTest.java:
```diff
@@ -74,13 +74,14 @@ void missingWorkspaceId() {
     CollectionId collectionId = CollectionId.of(UUID.randomUUID());
     // expect getWorkspaceId to throw, since it found an unexpected workspace id
-    assertThrows(RuntimeException.class, () -> collectionService.getWorkspaceId(collectionId));
+    assertThat(collectionService.getWorkspaceId(collectionId)).isEqualTo(mismatchingWorkspaceId);
```
There's been a lot of changes so apologies if I'm a bit behind.  I know multitenant will allow mismatching workpsace/collection IDs, but single tenant does not. Where are we enforcing the matching case for single tenant?

jladieu replied:
Timely that you should raise this question :)

Linking a comment attached to the stanza of removals where I'm pulling that enforcement out: https://github.com/DataBiosphere/terra-workspace-data-service/pull/602#discussion_r1508431660

I think we need to remove that safeguard if we want both of the following to be simultaneously true:

- Tests can setup/teardown random collections as part of fixture setup
- CollectionService has internally consistent implementation for canDeleteCollection (look up WorkspaceId based on CollectionId when figuring out which workspace to use for auth)

ashanhol commented on service/src/test/java/org/databiosphere/workspacedataservice/service/CollectionServiceSamExceptionTest.java:
```diff
@@ -38,19 +39,25 @@
- * RestException with a 500 error code.
+ * createCollection and deleteCollection to respond appropriately:
+ *
+ * <ul>
```
did this format automatically? if so, how did you get it to do that?

jladieu replied:
I wish...just toiled through it 😆

ashanhol commented on service/src/test/java/org/databiosphere/workspacedataservice/service/ImportServiceDataPlaneTest.java:
```diff
@@ -141,5 +131,14 @@ void collectionDoesNotExist() {
+    verifyNoInteractions(samAuthorizationDao);
+  }
+
+  private OngoingStubbing<Boolean> stubWriteWorkspacePermission(WorkspaceId workspaceId) {
```
:O i didn't know you could do this! That's so useful when you want some of the same whens but return different things in different cases

jladieu replied:
I know right?!  I discovered that as part of the big PR that preceded this one.  I like how it makes it easy to express multi-step stub chains as a nice, self-documenting single method call.

yuliadub commented on service/src/test/java/org/databiosphere/workspacedataservice/dao/MockCollectionDao.java:
```diff
@@ -71,7 +72,7 @@ public void alterSchema(UUID sourceWorkspaceId, UUID workspaceId) {
 
   @Override
   public WorkspaceId getWorkspaceId(CollectionId instanceId) {
```
just cause this caught my eye - can you rename to collectionId instead of instanceId? (or do we even need it?)

jladieu replied:
🦅 👁️

jladieu commented:
I put the safety check back in, put some tests back the way they were before I removed it, and added `tenancy.enforce-collections-match-workspace-id` to disable the check in stubborn test(s).

jladieu commented on service/src/test/java/org/databiosphere/workspacedataservice/tsv/TsvErrorMessageTest.java:
```diff
@@ -23,10 +23,16 @@
+@TestPropertySource(
+    properties = {
+      "twds.instance.workspace-id=123e4567-e89b-12d3-a456-426614174000",
+      "twds.tenancy.enforce-collections-match-workspace-id=false",
```
This is the only test where I ended up needing the killswitch.  I was a bit surprised by this though because I encountered a whole bunch of these the first time I tried to make this change.  This leads me to the suspicion that there is/was some kind of cross-test pollution, but I couldn't repro it on this go-around :shrug:

davidangb commented on service/src/main/resources/application.yml:
```diff
@@ -132,6 +132,7 @@ twds:
   tenancy:
     allow-virtual-collections: false
     require-env-workspace: true
+    enforce-collections-match-workspace-id: true
```
is it necessary to have this in both `application.yml` and `data-plane|control-plane.yml`? If the default is `true`, we could remove it from the data plane

jladieu replied:
I don't think it _should_ be, but for reasons of wanting to just get this merged, I am just gonna leave it alone.

Based on my comment in `application.yml` attached to `tenancy` and `data-import`, I think these are purposefully duplicated, although the fact that I felt like I needed to does seem to point to a problem we can look at on a rainy day...

davidangb reviewed (approved):
Minor comments inline, giving this a :+1:.

I don't like having the killswitch on the one test that needs it; that feels like we're doing something wrong. BUT, I also want to try to get away from refactoring/cleanup and focus on the cWDS MVP, so I am content to leave it as-is and save that for a future ticket.

jladieu commented:
FYI, I immediately encountered a problematic test after merging this, which leads me to think there's some kind of test pollution/ordering/seeding problem which necessitated introducing (and using) the kill switch from this PR: https://github.com/DataBiosphere/terra-workspace-data-service/pull/624
